### ISSUE TITLE
The results of reproducing this idea

### ISSUE THREAD

**Issue by Chen-Song** [xu-ji/IIC#20]
Great work!
I'll refer to your code and paper and reproduce it. However, the results were not satisfactory.
First, define three transformations in __init__ method:
        self.transform_tf1= transforms.Compose([transforms.RandomCrop(20),
                                    transforms.Resize(32),
                                    custom_greyscale_to_tensor(include_rgb=False)
                                    ])

        self.transform_tf2= transforms.Compose([transforms.RandomCrop(20),
                                    transforms.Resize(32),
                                    transforms.RandomHorizontalFlip(),
                                    transforms.ColorJitter(brightness=0.4, contrast=0.4,saturation=0.4, hue=0.125),
                                    custom_greyscale_to_tensor(include_rgb=False)
                                    ])

        self.transform_tf3=transforms.Compose([transforms.CenterCrop(20),
                                    transforms.Resize(32),
                                    custom_greyscale_to_tensor(include_rgb=False)
                               ])

Second, apply these transformations in __getitem__ method:

        img1, target = self.all_data[index], self.all_labels[index]
        img1 = Image.fromarray(img1)
        if not self.mapping_test_dataloader_flag:                 #build the head data
            img1_temp=[]
            img2_temp=[]
            img1_=self.transform_tf1(img1)
            for i in range(config.num_loaders):
                img1_temp.append(img1_)
                img2_temp.append(self.transform_tf2(img1))
            return img1_temp , img2_temp
        else:                                                                        #build the mapping data
            return self.transform_tf3(img1),target  

Third, all settings refer to your code, including optimizer, learning rate, etc. But in training stage, the best acc is low (on epoch 10, the best acc=15.14%), and the accuracy rate of each epoch increases very slowly(about 0.5%).

So did I ignore any important settings? Look forward to your reply.

**Comment by Chen-Song** [comment]
I'm sorry, it's my first time to write the issue and there is a problem with this format about self.transform_tf1.

    self.transform_tf1= transforms.Compose([transforms.RandomCrop(20),
                                  transforms.Resize(32),
                                  custom_greyscale_to_tensor(include_rgb=False)
                                ])
    self.transform_tf2= transforms.Compose([transforms.RandomCrop(20),
                                transforms.Resize(32),
                                transforms.RandomHorizontalFlip(),
                                transforms.ColorJitter(brightness=0.4, contrast=0.4,saturation=0.4, hue=0.125),
                                custom_greyscale_to_tensor(include_rgb=False)
                                ])

    self.transform_tf3=transforms.Compose([transforms.CenterCrop(20),
                                transforms.Resize(32),
                                custom_greyscale_to_tensor(include_rgb=False)
                           ])

**Comment by Chen-Song** [comment]
I did the experiment on CIFAR-10. Sorry to forget to mention this important information.

**Comment by xu-ji** [comment]
Hi, these questions come to mind based on your code snippets.

- Are you doing sobel filtering? 
- Also, auxiliary overclustering (two heads trained in alternate epochs)?
- How are you assessing accuracy, are you using linear assignment?
- If you are not using multiple sub-heads (multiple instances of last layer), you may need to change the learning rate, because our experiment used 5 sub-heads.
- Are you using the same architecture as we did?

**Comment by Chen-Song** [comment]
Hi ,I used your code in CIFAR-10 before (network architecture changed to Network In Network, head A and head B are both one sub-head, other things including learning rate, optimizer, sobel process are the same as your code), and the accuracy can reach 57% at about 140 epoch.
![plots](https://user-images.githubusercontent.com/38151697/63205086-fdc8bd80-c0d1-11e9-98ab-dbb59fae35a9.png)
But when I tried to reproduce the data transformer part, just write the transform_tf1,tf2,tf3 in ImageFolder. In my reproduce, the network architecture changed to Network In Network, head A and head B are both one sub-head, the lr=0.0001 and the sobel process, the overcluster are same with your code, but the experimental curves fluctuate greatly and can not converge.
![plots](https://user-images.githubusercontent.com/38151697/63205141-f48c2080-c0d2-11e9-9bc9-344f9de7f1d5.png)
You say it when there is one-subhead, I should modify the learning rate. I tried 0.001 and 0.00001, but the experimental curve also fluctuated greatly. So I feel a little confused about my reproduced transformer part. Thank you for your reply.

**Comment by xu-ji** [comment]
The loss is not steadily decreasing, which means there is certainly some problem.

Why are the loss values so large in magnitude? This is strange. You can see the difference as the top graph shows [0, -3] and the bottom shows [0, -60k]. They are the same dataset with same number of classes so this should not be so different. 

You changed the architecture. This is very important. (The learning rate would definitely need to change, possibly by a lot.) You could try using my original architecture instead of your network-in-network.

**Comment by Chen-Song** [comment]
Thank you for your advice. I'll try it.

**Comment by Chen-Song** [comment]
Hi, could you tell me why not do the 'demean' operation and the lr_schedule is empty list in CIFAR-10 experiment. In my experiemnts, I found the loss convergence about epoch=100, but in your setting, the max epoch=200, so why not to decline the lr. 
![plots](https://user-images.githubusercontent.com/38151697/63227866-fd3d3d80-c21d-11e9-99ff-518bf31499a0.png)

Most of deep learning experiments do the 'demean' operation, but why in CIFAR-10 Cluster experiment not?
Best.

**Comment by xu-ji** [comment]
The lr schedule option was something I added to the code but never really tested, because it adds extra complication, and we got the numbers we needed without it. But by all means you can try setting a lr schedule. 

Demeaning: from what I recall, we simply found the models performed better without demeaning.
